consensus/src/liveness/leader_reputation.rs

```diff
@@ -514,6 +474,59 @@ impl ProposerElection for LeaderReputation {
             .to_vec();
 
         let chosen_index = choose_index(weights, state);
-        self.proposers[chosen_index]
+        proposers[chosen_index]
     }
 }
+
+pub fn extract_epoch_to_proposers(
+    proof: EpochChangeProof,
+    epoch: u64,
+    proposers: &Vec<Author>,
+    first_epoch_to_consider: u64,
+    needed_rounds: u64,
+) -> Result<HashMap<u64, Vec<Author>>> {
+    ensure!(proof.ledger_info_with_sigs.len() as u64 == (epoch - (first_epoch_to_consider - 1)));
+    let last_index = proof.ledger_info_with_sigs.len() - 1;
+    let mut num_rounds = 0;
+    let mut result = HashMap::new();
+    for (index, ledger_info) in proof.ledger_info_with_sigs.into_iter().enumerate().rev() {
+        let cur_epoch_state = ledger_info
+            .ledger_info()
+            .next_epoch_state()
+            .ok_or_else(|| anyhow::anyhow!("no cur_epoch_state"))?;
+        let cur_epoch_proposers = cur_epoch_state
+            .verifier
+            .get_ordered_account_addresses_iter()
+            .collect::<Vec<_>>();
+        if index == last_index {
+            ensure!(
+                epoch == cur_epoch_state.epoch,
+                "fetched epoch_ending ledger_infos are for a wrong epoch {} vs {}",
+                epoch,
+                cur_epoch_state.epoch
+            );
+            ensure!(
+                proposers == &cur_epoch_proposers,
+                "proposers from state and fetched epoch_ending ledger_infos are missaligned"
+            );
+        }
+        result.insert(cur_epoch_state.epoch, cur_epoch_proposers);
+
+        if num_rounds > needed_rounds {
+            break;
+        }
+        // LI contains validator set for next epoch (i.e. in next_epoch_state)
+        // so after adding the emount of rounds in the epoch, we need to process the
+        // corresponding Validatorset on the previous ledger info,
+        // before checking if we are done.
+        num_rounds += ledger_info.ledger_info().round();
```

Comment:
I think we don't expect to go beyond more than 1 epoch, does it make sense to simplify it?

Reply:
Well, this was @sitalkedia's comment on the previous diff.

I thought to argue against it, but if epoch can be triggered by reconfiguration, then epoch can be very short. And we don't want network to start behaving badly in that case. 
So this is not too complicated, and has a limit (use_history_from_previous_epoch_max_count) which if set to 1, has the equivalent meaning.